### ISSUE TITLE
Best practice for training large scale dataset like imagenet using Lasagne

### ISSUE THREAD

**Issue by kli-nlpr** [Lasagne/Lasagne#615]
Hey guys, I found that all of the examples of Lasagne deal with small data set like mnist and cifar10 which can be loaded into memory completely.
My question is how to write efficient code for training large scale dataset?
Specifically, what is the best way to prepare mini-batches (including real time data augmentation) in order to keep the GPU busy?

Thank you very much!

**Comment by f0k** [comment]
> My question is how to write efficient code for training on large scale datasets?

The MNIST example already sports an `iterate_minibatches()` function, a generator function that returns one mini-batch at a time. You can use the same scheme for iterating over files that are *not* yet loaded into memory. For example, you could dump your data into multiple files that are then accessed in random order, or into a single large `.npy` file that is loaded with `np.load(..., mmap_mode='r')` to be read as a memory-mapped file (this would allow you to use the same iteration code as before, since it can be treated like a regular numpy array).

> Specifically, what is the best way to prepare mini-batches (including real time data augmentation) in order to keep the GPU busy?

To do so, you can run your generator function in a background thread: https://github.com/Lasagne/Lasagne/issues/12#issuecomment-59494251
Theano will release the GIL when waiting for GPU operations to finish, so this allows your generator function to be run in parallel.
Your generator function can include data augmentation as well. Often you will find it more convenient to stack multiple generator functions -- one for reading the data, one for performing agumentation X, one for performing augmentation Y, ... I've got example code of that structure for real-time data augmentation of audio spectrograms here: http://ofai.at/~jan.schlueter/code/augment/

Important caveat: If your data is stored in HDF5 format, any file I/O will take the GIL again, which will reduce the amount of parallelism a lot. You will need to resort to the `multiprocessing` module then (requires minimal changes to my background thread example). This, in turn, will incur some overhead for transferring the data from one process to another via pickling and unpickling it. `joblib` includes a workaround for numpy arrays that resorts to memory-mapped files for faster transfer.

Hope this helps!

**Comment by kli-nlpr** [comment]
f0k,  Thank you very much  o(^▽^)o

**Comment by kli-nlpr** [comment]
By the way, It is definitely helpful if anyone can write a ipython notebook example about training on large scale data set. (For example, like CAFFE's image data layer that read images from a large txt file which contain image path and labels.)

**Comment by ebenolson** [comment]
I've found I get better throughput for loading + data augmentation by using processes and writing results as pickles in shared memory (`/run/shm` on Ubuntu). I think that `joblib` is optimized for the opposite case of distributing large inputs and receiving small results back.

I have a gist showing the general approach I've used here: https://gist.github.com/ebenolson/072712792c46aa192797

**Comment by kli-nlpr** [comment]
Maybe use the method which keras provided?
http://keras.io/faq/#how-can-i-use-keras-with-datasets-that-dont-fit-in-memory
I will check out this.
Thank you very much, guys~

**Comment by chuzui** [comment]
I think [theano_alexnet](https://github.com/uoguelph-mlrg/theano_alexnet)  is useful. It uses two processes to train and preprocess data, ZMQ to exchange message between two processes and PyCuda to convert GPU memory handle. You can easily write a Lasagne example like it

**Comment by f0k** [comment]
> I think theano_alexnet is useful.

Interesting! They create a PyCUDA variable from a Theano shared variable, obtain an IPC handle to its GPU memory, and then have the second process write its preprocessed batches directly to GPU memory (which is large enough to house multiple batches, like a circular buffer). This way the second process doesn't have to pass the data to the first process as a numpy array. Would be even cooler if Theano would allow using a non-null CUDA stream, so the second process could write to GPU while the first process computes on it (I think that part is still forced to be sequential now). Thanks for the pointer!

**Comment by DeanChan** [comment]
> Specifically, what is the best way to prepare mini-batches (including real time data augmentation) in order to keep the GPU busy?

```python
from numba import jit

@jit(nogil=True)
def feed_mini_batch():
...
```

This may be more user-friendly than using Threading.

**Comment by f0k** [comment]
> This may be more user-friendly than using Threading.

You still need threads. The main thread waits for the theano function call to finish, so you need a second thread that can call your `feed_mini_batch()` routine in the meantime. Freeing the GIL inside `feed_mini_batch()` would allow a *third* interpreter thread to do something while the first thread waits for the theano function and the second thread waits for the `feed_mini_batch()` routine to return/yield something. Or am I missing anything?
